minetest/minetest
# Add dynamic exposure correction

This PR adds shader code to simulate behavior of human eye and make proper use of the bloom.

# Process

1. Average scene color and luminance are computed using progressively downsampled screen image. Color is sampled at `(x,y) = (0.5, 0.5) + (0.5, 0.5) * pow((i, j), center_weight_power)` for i, j from 0 to 1.
2.  Target exposure is calculated according to the [Frostbite](https://media.contentapi.ea.com/content/dam/eacom/frostbite/files/course-notes-moving-frostbite-to-pbr-v2.pdf) paper.
3. Exposure is blended in time using exponential falloff `current_exposure = previous_exposure + (target_exposure - previous_exposure) * (1 - exp(-speed * time)` to untie the speed of transition from FPS.
4. Final color value is `color = color * pow(2, current_exposure) * compensation_factor`

# Client-side controls

* `enable_auto_exposure` setting allows to enable the feature on the client.

# Game-side controls

 there are new parameters to `player:set_lighting`:

```
player:set_lighting({
  exposure = {
    luminance_min = <float>, -- lower bound of observed scene luminance
    luminance_max = <float>, -- upper bound of observed scene luminance
    luminance_bias = <float>, -- bias value in the equation. Higher values make scene darker
    luminance_key = <float>, -- target luminance value in the equation
    speed_dark_bright = <float>, -- speed of adaptation to bright light
    speed_bright_dark = <float>, -- speed of adaptation to dark scene
    center_weight_power = <float>, -- power for center-weighting the color samples
    compensation_factor = <float>, -- final compensation factor
  }
})
```

## To do

This PR is Ready for Review.

## How to test

1.Set `enable_auto_exposure` to true. Optionally set `enable_bloom` to true.
2. Start devtest
3. Walk in bright and dark areas and observe how scene brightness is corrected automatically, and the amount of e.g. bloom changes.
4. Type `/set_lighting`
5. Play with the controls on the right side of the screen - every control must have effect on the screen.
6. For other games there is an example [mod](https://github.com/x2048/dynamic_exposure/).
